src/MacroTools/Mechanics/StructurePacking.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections.Generic;
+using WCSharp.Events;
+using static War3Api.Common;
+
+namespace MacroTools.Mechanics
+{
+  /// <summary>
+  /// System to allow structures to convert themselves into units. Those units can then use an ability to rebuild the building.
+  /// </summary>
+  public class StructurePacking
+  {
+    /// <summary>
+    /// Object that contains information and methods for packable structures
+    /// </summary>
+    public class PackableStructure
+    {
+      private static readonly Dictionary<int, PackableStructure> PackableStructureById = new();
+      private int _packedUnitId;
+      private int _buildAbility;
+      private string? _structureModel;
+
+      /// <summary>
+      /// Method for creating new packable structure entries
+      /// </summary>
+      /// <param name="structureId"></param>
+      /// <param name="packedUnitId"></param>
+      /// <param name="buildAbility"></param>
+      /// <param name="structureModel"></param>
+      public static void Create(int structureId, int packedUnitId, int buildAbility, string? structureModel)
+      {
+        var packable = new PackableStructure
+        {
+          _packedUnitId = packedUnitId,
+          _buildAbility = buildAbility,
+          _structureModel = structureModel
+        };
+        PackableStructureById[structureId] = packable;
+        
+        PlayerUnitEvents.Register(PlayerUnitEvent.UnitTypeFinishesTraining, GetPackableStructureById(structureId).OnTrainUnitType,structureId);
```

Comment:
So I suspect we can simplify this whole trigger massively through the use of anonymous functions. We didn't have those in the vJASS version you looked at so we had to use the Dictionary technique. Let's chat about it on Discord some time.